File: Models/TreeNodeOfT.vb
Imports System
Imports System.ComponentModel
Imports System.Drawing
Imports BrightIdeasSoftware
Imports System.Collections
Imports System.IO
Imports Microsoft.Office.Interop.Outlook
Imports System.Collections.Generic
Imports System.Linq
Imports System.Collections.ObjectModel
Imports System.Diagnostics
Imports Microsoft.Office.Core
Imports System.Linq.Expressions

Public Class TreeNode(Of T)
    Private _Parent As TreeNode(Of T)
    Private ReadOnly _value As T
    'Private ReadOnly _children As List(Of TreeNode(Of T)) = New List(Of TreeNode(Of T))()
    Private _children As List(Of TreeNode(Of T)) = New List(Of TreeNode(Of T))()
    Public ID As String

    Public Sub New(ByVal value As T)
        _value = value
    End Sub

    Public Sub New(ByVal value As T, ByVal strID As String)
        _value = value
        ID = strID
    End Sub

    Default Public ReadOnly Property Item(ByVal i As Integer) As TreeNode(Of T)
        Get
            Return _children(i)
        End Get
    End Property

    Public ReadOnly Property ParentID As String
        Get
            If _Parent Is Nothing Then
                Return "Root"
            Else
                Return _Parent.ID
            End If
        End Get
    End Property
    Public Property Parent As TreeNode(Of T)
        Get
            Return _Parent
        End Get
        Set(value As TreeNode(Of T))
            _Parent = value
        End Set
    End Property

    Public ReadOnly Property Value As T
        Get
            Return _value
        End Get
    End Property

    Public Function IsAncestor(ByVal model As TreeNode(Of T)) As Boolean
        If Me Is model Then Return True
        If _Parent Is Nothing Then Return False
        Return _Parent.IsAncestor(model)
    End Function

    Public ReadOnly Property ChildCount As Integer
        Get
            Return _children.Count
        End Get
    End Property

    'Public ReadOnly Property NextChildId As String
    '    Get
    Public Function NextChildID() As String
        Dim strMaxID As String = ID & "00"
        Dim lngMaxID As Long = ConvertToDecimal(125, strMaxID)
        Dim strTmpID As String = ""
        Dim lngTmpID As Long = 0
        For Each child In Children
            strTmpID = child.ID
            lngTmpID = ConvertToDecimal(125, strTmpID)
            If lngTmpID > lngMaxID Then
                lngMaxID = lngTmpID
            End If
        Next child

        Dim blContinue As Boolean = True
        While blContinue
            lngMaxID += 1
            strMaxID = ConvertToBase(125, lngMaxID)
            If Globals.ThisAddIn.UsedIDList.Contains(strMaxID) = False Then
                blContinue = False
            End If
        End While
        Globals.ThisAddIn.UsedIDList_Append(strMaxID)
        Return strMaxID
    End Function


    Public Property Children As List(Of TreeNode(Of T))
        Get
            Return _children
        End Get
        Set(value As List(Of TreeNode(Of T)))
            _children = value
        End Set
    End Property

    'Public ReadOnly Property Children As ReadOnlyCollection(Of TreeNode(Of T))
    '    Get
    '        Return _children.AsReadOnly()
    '    End Get
    'End Property

    Public Function AddChild(ByVal value As T) As TreeNode(Of T)
        Dim node = New TreeNode(Of T)(value) With {
            .Parent = Me
            }
        node.ID = NextChildID()
        _children.Add(node)
        Return node
    End Function
    Public Function AddChild(ByVal node As TreeNode(Of T)) As TreeNode(Of T)
        'node.Parent = Me
        node.ID = NextChildID()
        _children.Add(node)
        Return node
    End Function
    Public Function AddChild(ByVal node As TreeNode(Of T), ByVal strID As String) As TreeNode(Of T)
        node.Parent = Me
        node.ID = strID
        _children.Add(node)
        Return node
    End Function
    Public Function InsertChild(ByVal node As TreeNode(Of T), ByVal strID As String) As TreeNode(Of T)
        node.Parent = Me
        node.ID = strID
        _children.Insert(0, node)
        Return node
    End Function
    Public Function AddChild(ByVal value As T, ByVal strID As String) As TreeNode(Of T)
        Dim node = New TreeNode(Of T)(value) With {
                .Parent = Me
            }
        node.ID = strID
        _children.Add(node)
        Return node
    End Function
    Public Function AddChildren(ParamArray values As T()) As TreeNode(Of T)()
        Return values.[Select](New Func(Of T, TreeNode(Of T))(AddressOf AddChild)).ToArray()
    End Function

    Public Function RemoveChild(ByVal node As TreeNode(Of T)) As Boolean
        Return _children.Remove(node)
    End Function
    'Public Sub SubsituteIDPrefix(ByVal strOld, ByVal strNew)
    '    If Mid(ID, 1, strOld.Length) = strOld Then
    '        ID = strNew & Mid(ID, strOld.Length + 1, ID.Length - strOld.Length)
    '    End If

    '    For Each child In _children
    '        child.SubsituteIDPrefix(strOld, strNew)
    '    Next

    'End Sub
    Public Sub Traverse(ByVal action As Action(Of T))
        action(Value)

        For Each child In _children
            child.Traverse(action)
        Next
    End Sub

    Public Function FindChildByID(strID As String) As TreeNode(Of T)
        Dim node As TreeNode(Of T)
        Dim rnode As TreeNode(Of T)
        If ID = strID Then
            Return Me
        Else
            For Each node In Children
                rnode = node.FindChildByID(strID)
                If Not rnode Is Nothing Then
                    Return rnode
                End If
            Next
            Return Nothing
        End If
    End Function

    Public Function Flatten() As IEnumerable(Of T)
        Return {Value}.Concat(_children.SelectMany(Function(x) x.Flatten()))
    End Function
End Class

File: UtilitiesCS/ReusableTypeClasses/Other/TreeNodeOfT.cs
using System;
using System.Collections.Generic;
using System.Data;
using System.Linq;
using UtilitiesCS.Extensions;

namespace UtilitiesCS
{
    public class TreeNode<T>
    {
        public TreeNode(T value) => Value = value;
        public TreeNode(TreeNode<T> node)
        {
            Parent = node.Parent;
            Value = node.Value;
            Children = node.Children;
        }

        #region Public Properties

        private TreeNode<T> _parent;
        public TreeNode<T> Parent { get => _parent; set => _parent = value; }

        private T _value;
        public T Value { get => _value; private set => _value = value; }

        private List<TreeNode<T>> _children = new List<TreeNode<T>>();
        public List<TreeNode<T>> Children { get => _children; set => _children = value; }
        public TreeNode<T> this[int i]
        {
            get
            {
                return Children[i];
            }
        }

        public int ChildCount { get => Children.Count; }

        public int Depth
        {
            get
            {
                if (Parent is null)
                    return 0;
                return Parent.Depth + 1;
            }
        }

        #endregion Public Properties

        #region Change Structure

        public TreeNode<T> AddChild(T value)
        {
            var node = new TreeNode<T>(value) { Parent = this };
            Children.Add(node);
            return node;
        }
        
        public TreeNode<T> AddChild(T value, string strID)
        {
            var node = new TreeNode<T>(value) { Parent = this };
            // node.ID = strID
            Children.Add(node);
            return node;
        }

        public TreeNode<T> AddChild(TreeNode<T> node)
        {
            // node.Parent = Me
            // node.ID = NextChildID()
            Children.Add(node);
            return node;
        }
        
        public TreeNode<T>[] AddChildren(params T[] values)
        {
            return values.Select(new Func<T, TreeNode<T>>(AddChild)).ToArray();
        }

        public TreeNode<T> InsertChild(TreeNode<T> node)
        {
            node.Parent = this;
            Children.Insert(0, node);
            return node;
        }

        public bool RemoveChild(TreeNode<T> node)
        {
            return Children.Remove(node);
        }

        #endregion Change Structure

        #region Public Methods

        public IEnumerable<TreeNode<T>> Descendents(bool includeSelf = false)
        {
            TreeNode<T>[] nodes = includeSelf ?[this]: [];
            return nodes.Concat(Children.SelectMany(x => x.Descendents(true)));
        }

        public TreeNode<T> FirstAncestor(Func<T, bool> condition)
        {
            if (Parent is null)
                return default;
            if (condition(Parent.Value))
                return Parent;
            return Parent.FirstAncestor(condition);
        }

        public object FindByDelegate(Func<T, string, bool> comparator, string StringToCompare)
        {
            foreach (var node in Children)
            {
                if (comparator(Value, StringToCompare))
                {
                    return node;
                }
            }
            return null;
        }

        public object FindByDelegate(Func<T, T, bool> comparator, T T2)
        {

            foreach (var node in Children)
            {
                if (comparator(Value, T2))
                {
                    return node;
                }
            }
            return null;
        }

        public TreeNode<T> FindSequentialNode<U>(Func<T, U, bool> comparator, Queue<U> sequence)
        {
            comparator.ThrowIfNull();
            var first = sequence.ThrowIfNullOrEmpty().Dequeue();
            var node = FindNode((current) => comparator(current, first), true);
            while (node is not null && !sequence.IsEmpty())
            {
                var next = sequence.Dequeue();
                node = node.Children?.Where(x => comparator(x.Value, next))?.FirstOrDefault();
            }
            return node;
        }

        public TreeNode<T> FindNode(Func<T, bool> comparator, bool descendByLevel = false) 
        { 
            if (!descendByLevel) 
            { 
                return FindAll(comparator).FirstOrDefault(); 
            }
            else
            {
                TreeNode<T>[] nodes = { this };
                while (!nodes.IsNullOrEmpty())
                {
                    var first = nodes.FirstOrDefault(node => comparator(node.Value));
                    if (first != default) { return first; }
                    else { nodes = GetNextLevel(nodes).Where(node => node is not null).ToArray(); }
                }
                return default;
            }
            
        }
        
        public TreeNode<T>[] GetNextLevel(TreeNode<T>[] nodes) 
        { 
            if (nodes is null) { return null; }
            return nodes.Where(x => !x.Children.IsNullOrEmpty()).SelectMany(x => x.Children).ToArray();
        }

        public IEnumerable<TreeNode<T>> FindAll(Func<T, bool> comparator)
        {
            if (comparator(this.Value))
                return new TreeNode<T>[] { this }.Concat(Children.SelectMany(x => x.FindAll(comparator)));
            else
                return new TreeNode<T>[] { }.Concat(Children.SelectMany(x => x.FindAll(comparator)));
        }

        public IEnumerable<TreeNode<T>> FindAll(Func<TreeNode<T>, bool> comparator) 
        {
            if (comparator(this))
                return new TreeNode<T>[] { this }.Concat(Children.SelectMany(x => x.FindAll(comparator)));
            else
                return new TreeNode<T>[] { }.Concat(Children.SelectMany(x => x.FindAll(comparator)));
        }
        
        public IEnumerable<T> Flatten()
        {
            return new[] { Value }.Concat(Children.SelectMany(x => x.Flatten()));
        }

        public IEnumerable<T> FlattenIf(Func<T, bool> comparator)
        {
            return Flatten().Where(comparator);
        }
        
        public bool IsAncestor(TreeNode<T> model)
        {
            if (ReferenceEquals(this, model))
                return true;
            if (Parent is null)
                return false;
            return Parent.IsAncestor(model);
        }
        
        public virtual IEnumerable<TreeNode<T>> Leaves()
        {
            TreeNode<T>[] nodes = [];
            return nodes.Concat(Children.SelectMany(x => 
            {
                if (x.Children.Count == 0)
                    return new[] { x };
                else
                    return x.Leaves();
            }));
        }
                
        public void Traverse(Action<T> action)
        {
            action(Value);

            foreach (var child in Children)
                child.Traverse(action);
        }

        public void Traverse(Action<TreeNode<T>> action)
        {
            action(this);

            foreach (var child in Children)
                child.Traverse(action);
        }

        #endregion Public Methods
    }
}